### PR TITLE
feat: replace overkill `getDMMF` with `validate`

### LINKED ISSUE #17618: Figure out where getDMMF is used for validation
getDMMF is much heavier than just validation, we would benefit from doing just validation in these places. This issue is only about investigating _where_ we are currently doing that in our whole typescript codebase.

Output: Notion document with overview.

### PR DESCRIPTION
**Pull request by jkomyno**
Closes https://github.com/prisma/prisma/issues/17618.

This replacement preserves the same visual output than before (coloring included), i.e., this is only an internal change.

----

Example usage of `validate` as a substitute of `getDMMF` in `prisma format`

(with colors):

Screenshot 2023-02-21 at 12 50 09

(without colors):

Screenshot 2023-02-21 at 12 50 34

### CODE REVIEW

File: packages/internals/src/__tests__/engine-commands/validate.test.ts [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+          Validation Error Count: 1
+          [Context: getDmmf]
+
+          Prisma CLI Version : 0.0.0"
+        `)
+      }
+    })
+
+    // TODO: wait for https://github.com/prisma/prisma/pull/17950 to be merged
+    test.skip(`panics when the given datamodel isnt' a string`, () => {
```
Comment (jkomyno) [inline packages/internals/src/__tests__/engine-commands/validate.test.ts:175]: Waiting for https://github.com/prisma/prisma-engines/pull/3714

**Comment by janpio** [comment]
Seems some error messages are too raw:

```
[2023-02-17T22:23:48Z]     - Snapshot  - 11
[2023-02-17T22:23:48Z]     + Received  +  2
[2023-02-17T22:23:48Z]
[2023-02-17T22:23:48Z]     - "Prisma schema validation - (validate wasm)
[2023-02-17T22:23:48Z]     - Error code: P1012
[2023-02-17T22:23:48Z]     - error: Error parsing attribute "@default": The `autoincrement()` default value is used on a non-indexed field even though the datasource does not support this.
[2023-02-17T22:23:48Z]     -   -->  schema.prisma:7
[2023-02-17T22:23:48Z]     -    | 
[2023-02-17T22:23:48Z]     -  6 |         model User {
[2023-02-17T22:23:48Z]     -  7 |           id        Int      @default(autoincrement())
[2023-02-17T22:23:48Z]     -  8 |           email     String   @unique
[2023-02-17T22:23:48Z]     -    | 
[2023-02-17T22:23:48Z]     -
[2023-02-17T22:23:48Z]     - Validation Error Count: 1
[2023-02-17T22:23:48Z]     + "(validate wasm)
[2023-02-17T22:23:48Z]     + Details: Error: {"error_code":"P1012","message":"\u001b[1;91merror\u001b[0m: \u001b[1mError parsing attribute \"@default\": The `autoincrement()` default value is used on a non-indexed field even though the datasource does not support this.\u001b[0m\n  \u001b[1;94m-->\u001b[0m  \u001b[4mschema.prisma:7\u001b[0m\n\u001b[1;94m   | \u001b[0m\n\u001b[1;94m 6 | \u001b[0m        model User {\n\u001b[1;94m 7 | \u001b[0m          \u001b[1;91mid        Int      @default(autoincrement())\u001b[0m\n\u001b[1;94m 8 | \u001b[0m          email     String   @unique\n\u001b[1;94m   | \u001b[0m\n\nValidation Error Count: 1"}
```

**Comment by jkomyno** [comment]
I'm really confused about why the tests still fail, considering that the Rust engines should be aligned already.

File: packages/internals/src/engine-commands/validate.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  /**
+   * Check which error to throw.
+   */
+  const error = match(validateEither.left)
+    .with({ type: 'wasm-error' }, (e) => {
+      debugErrorType(e)
+
+      console.error('') // empty line for better readability
```
Comment (jkomyno) [inline packages/internals/src/engine-commands/validate.ts:87]: This was copied from the catch block of `Format.ts` and `Validate.ts`

File: packages/cli/src/__tests__/artificial-panic.test.ts [hunk 3]
```diff
@@ -134,14 +134,14 @@ describe('artificial-panic get-config', () => {
   })
 })
 
-describe('artificial-panic get-dmmf', () => {
+describe('artificial-panic validate', () => {
```
Comment (millsp) [inline packages/cli/src/__tests__/artificial-panic.test.ts:137]: Do we also have an artificial panic test for getDMMF somewhere else?

File: packages/cli/src/__tests__/commands/Validate.test.ts [hunk 4]
```diff
@@ -39,9 +39,9 @@ describe('validate', () => {
     // stderr
     expect(ctx.mocked['console.warn'].mock.calls.join('\n')).toMatchInlineSnapshot(`
 
-                              Prisma schema warning:
-                              - Preview feature "nativeTypes" is deprecated. The functionality can be used without specifying it as a preview feature.
-                  `)
+                                    Prisma schema warning:
```
Comment (millsp) [inline packages/cli/src/__tests__/commands/Validate.test.ts:42]: Do you know why the indent is slightly changed?

File: packages/cli/src/__tests__/incomplete-schemas.test.ts [hunk 5]
```diff
@@ -30,10 +30,15 @@ DROP TABLE IF EXISTS 'test-dbexecute';
 CREATE TABLE 'test-dbexecute' ("id" INTEGER PRIMARY KEY);
 DROP TABLE 'test-dbexecute';`
 
-function urlIsMissingValidationError(source: 'getDmmf' | 'getConfig') {
-  const header = source === 'getDmmf' ? 'get-dmmf wasm' : 'get-config wasm'
+type Source = 'get-config' | 'validate'
+const sourceToContext = {
+  'get-config': 'getConfig',
```
Comment (millsp) [inline packages/cli/src/__tests__/incomplete-schemas.test.ts:35]: Why does `getConfig` becomes `get-config` for wasm?

File: packages/internals/src/engine-commands/validate.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+         */
+        if (process.env.FORCE_PANIC_QUERY_ENGINE_GET_DMMF) {
+          debug('Triggering a Rust panic...')
+          prismaFmt.debug_panic()
+        }
+
+        const noColor = Boolean(process.env.NO_COLOR)
+        const params = JSON.stringify({
+          prismaSchema: options.datamodel,
+          noColor,
```
Comment (millsp) [inline packages/internals/src/engine-commands/validate.ts:63]:
```suggestion
        const params = JSON.stringify({
          prismaSchema: options.datamodel,
          noColor: Boolean(process.env.NO_COLOR),
```

File: packages/internals/src/engine-commands/validate.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+export class ValidateError extends Error {
+  constructor(params: QueryEngineErrorInit) {
+    const constructedErrorMessage = match(params)
+      .with({ _tag: 'parsed' }, ({ errorCode, message, reason }) => {
+        const errorCodeMessage = errorCode ? `Error code: ${errorCode}` : ''
+        return `${reason}
+${errorCodeMessage}
+${message}`
+      })
+      .with({ _tag: 'unparsed' }, ({ message, reason }) => {
```
Comment (millsp) [inline packages/internals/src/engine-commands/validate.ts:27]: In which case does this happen?

File: packages/cli/src/__tests__/commands/Validate.test.ts (hunk 4 above)
Comment (jkomyno) [reply packages/cli/src/__tests__/commands/Validate.test.ts:42]: It's just `jest` with the `-u` flag, it's almost never consistent although the warnings didn't change at all

File: packages/internals/src/engine-commands/validate.ts (hunk 7 above)
Comment (jkomyno) [reply packages/internals/src/engine-commands/validate.ts:27]: When the error message (which we assume being a `JSON`-encoded string) isn't parseable via `JSON.parse`.
See [here](https://github.com/prisma/prisma/blob/bfbd3ec077596e4554f56ac20a2cde663cef8b3b/packages/internals/src/engine-commands/queryEngineCommons.ts#L131-L134).

File: packages/cli/src/__tests__/artificial-panic.test.ts (hunk 3 above)
Comment (jkomyno) [reply packages/cli/src/__tests__/artificial-panic.test.ts:137]: You're right that with that test replacement, the artificial panic logic for `getDMMF` was no longer tested. I've added it separately now.

**Review by millsp — APPROVED** [review]

**Pull request merged**